flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java

```diff
@@ -877,4 +899,610 @@ private String getSensorPositionInfo(Asset asset) {
+            }
+            break;
+         case MetricName.PDU_VOLTAGE:
+            if(extraidentifier != null) {
+               String extraidentifierList[] = extraidentifier.split("\\|");
+               if(extraidentifierList.length == 1) {
+                  metricData.setMetricName(
+                        String.format(MetricKeyName.PDU_XLET_VOLTAGE, extraidentifier));
+               }else if(extraidentifierList.length == 2) {
+                  String inlet = extraidentifierList[0];
+                  String pole = extraidentifierList[1];
+                  metricData.setMetricName(
+                        String.format(MetricKeyName.PDU_INLET_POLE_VOLTAGE, inlet, pole));
+               }
+               result.add(metricData);
+            }
+            break;
+         case MetricName.PDU_FREE_CAPACITY:
+            if(extraidentifier != null) {
+               String extraidentifierList[] = extraidentifier.split("\\|");
+               if(extraidentifierList.length == 1) {
+                  metricData.setMetricName(String.format(MetricKeyName.PDU_XLET_FREE_CAPACITY,
+                        extraidentifier));
+               }else if(extraidentifierList.length == 2) {
+                  String inlet = extraidentifierList[0];
+                  String pole = extraidentifierList[1];
+                  metricData.setMetricName(
+                        String.format(MetricKeyName.PDU_INLET_POLE_FREE_CAPACITY, inlet, pole));
+               }
+               result.add(metricData);
+            }
+            break;
+         default:
+            break;
+         }
+      }
+      return result;
+   }
+
+   private Asset getAssetById(String assetID) {
+      Optional<Asset> assetOptional = assetRepository.findById(assetID);
+      if (!assetOptional.isPresent()) {
+         throw WormholeRequestException.NotFound("asset", "id", assetID);
+      }
+      return assetOptional.get();
+   }
+
+   private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<String> metricsName) {
+      Iterator<ValueUnit> ite = valueunits.iterator();
+      while (ite.hasNext()) {
+         ValueUnit valueUnit = ite.next();
+         if(!metricsName.contains(valueUnit.getKey())) {
+            ite.remove();
+         }
+      }
+   }
+
+   private List<ValueUnit> getValueUnitsByID(String assetID, long starttime, int duration) {
+      List<RealTimeData> realtimeDatas =
+            realtimeDataRepository.getDataByIDAndTimeRange(assetID, starttime, duration);
+      List<ValueUnit> valueunits = new ArrayList<ValueUnit>();
+      if(realtimeDatas.isEmpty()) {
+         return valueunits;
+      }
+      for(RealTimeData realTimeData : realtimeDatas) {
+         for(ValueUnit value : realTimeData.getValues()) {
+            valueunits.add(value);
+         }
+      }
+      return valueunits;
+   }
+
+   private Set<String> getSensorAssetIDFromFormula(Map<String, Map<String, String>> sensorFormulasMap){
```

Comment:
Where does the formula come from? Server or PUD or both?
if both extractSensorIDfromFormula

otherwise should be  
extractSensorIDfromServerFormula

extractSensorIDfromPDUFormula